<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8'>
    <meta http-equiv="X-UA-Compatible" content="chrome=1">
    <meta name="description" content="IVA : Iterative Virus Assembler">

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>IVA</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/sanger-pathogens/iva">View on GitHub</a>

          <h1 id="project_title">IVA</h1>
          <h2 id="project_tagline">Iterative Virus Assembler</h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/sanger-pathogens/iva/archive/v1.0.3.zip">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/sanger-pathogens/iva/archive/v1.0.3.tar.gz">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <h1>
<a name="overview" class="anchor" href="#overview"><span class="octicon octicon-link"></span></a>Overview</h1>

<p>IVA is a <em>de novo</em> assembler designed to assemble virus genomes that
have no repeat sequences, using Illumina read pairs sequenced from
mixed populations at extremely high and variable depth.</p>

<p>For more information, please read the <a href="http://bioinformatics.oxfordjournals.org/content/early/2015/02/27/bioinformatics.btv120.abstract">IVA publication</a> (open access).</p>

<p>Installation instructions are below. For usage help and examples, see
the <a href="https://github.com/sanger-pathogens/iva/wiki">IVA wiki page</a>.</p>

<hr><h1>
<a name="installation" class="anchor" href="#installation"><span class="octicon octicon-link"></span></a>Installation</h1>
<p>IVA was developed for and intended to be run on Linux. It has also
been run successfully on OS X. If you have a Windows machine 
(or even Linux or a Mac) then you can run IVA using a virtual machine with
<a href="https://www.virtualbox.org/">VirtualBox</a>.

<p>General installation instructions for Linux/OS X are below. You may also find it useful to read
  <a href="https://github.com/sanger-pathogens/iva/wiki/Installation-notes-for-Ubuntu">detailed instructions for Ubuntu</a>.
</p>

<h4>
<a name="required-dependencies" class="anchor" href="#required-dependencies"><span class="octicon octicon-link"></span></a>Required dependencies</h4>

<p>The following are required to install IVA and use it to run an assembly.</p>

<ul>
<li>
<a href="http://www.python.org/">Python 3</a> version 3.3 or higher (IVA is written in Python 3)
</li>
<li>
<a href="http://sun.aei.polsl.pl/kmc/download.html">KMC</a> installed, so that <code>kmc</code> and <code>kmc_dump</code> are in your path.</li>
<li>
<a href="http://mummer.sourceforge.net/">MUMmer</a> installed with its executables (ie <code>nucmer</code> etc)
in your path.</li>
<li>
<a href="http://samtools.sourceforge.net/">Samtools</a> installed, so that <code>samtools</code> is in your path.</li>
<li>
<a href="http://www.sanger.ac.uk/resources/software/smalt/">SMALT</a> installed, so that <code>smalt</code> is in your path.</li>
<li>Optional: <a href="http://www.usadellab.org/cms/?page=trimmomatic">Trimmomatic</a> - although this is optional, it is
highly recommended.
It is used to trim adapter sequences from reads before assembling and
significantly improves the results.
You don't need to add anything to your path, but will
need to tell IVA where the Java jar file is to use Trimmomatic (see
<a href="https://github.com/sanger-pathogens/iva/wiki/iva-examples">examples</a>).</li>
</ul><p>The recommended versions are: kmc version 2.1.1, MUMmer version 3.23, samtools
version 0.1.19 or greater, and SMALT version 0.7.6.</p>

<h4>
<a name="install-iva" class="anchor" href="#install-iva"><span class="octicon octicon-link"></span></a>Install IVA</h4>

<p>Once you have installed the dependencies, install IVA with

<pre><code>pip3 install iva
</code></pre>
</p>

<p>The installation can be tested by
<a href="https://github.com/sanger-pathogens/iva/wiki/Assembly-test-data">running an assembly using test data included with IVA</a>.
</p>

<h4>
<a name="Docker" class="anchor" href="#docker"><span class="octicon octicon-link"></span></a>Docker</h4>
<p>IVA can be run in a Docker container. First install Docker, then install IVA: </p>

<pre><code>docker pull sangerpathogens/iva
</pre></code>
<p>To use it you would use a command such as this (substituting in your directories), where your files are assumed to be stored in /home/ubuntu/data: </p>

<pre><code>docker run --rm -it -v /home/ubuntu/data:/data sangerpathogens/iva iva -f /data/reads_fwd.fastq -r /data/reads_rev.fastq /data/Output_directory
</pre></code>

<p>The image includes all required and optional dependencies for IVA. To run IVA with trimmomatic, you also need to specify the path to the trimmomatic jar file, which is available at /Trimmomatic-0.38/trimmomatic-0.38.jar:</p>
<pre><code>docker run --rm -it -v /home/ubuntu/data:/data sangerpathogens/iva iva --trimmomatic /Trimmomatic-0.38/trimmomatic-0.38.jar -f /data/reads_fwd.fastq -r /data/reads_rev.fastq /data/Output_directory
</pre></code>

<h4>
<a name="optional-dependencies" class="anchor" href="#optional-dependencies"><span class="octicon octicon-link"></span></a>Optional dependencies</h4>

<p>The following dependencies only apply to the QC scripts of IVA and
are not needed to run an assembly. Only install them if you want to run the scripts
iva_qc or iva_qc_make_db.
</p>

<ul>
<li>
<a href="http://www.r-project.org/">R</a> installed and in your path.</li>
<li><a href="http://www.bioperl.org/wiki/Main_Page">BioPerl</a></li>
<li>Optional: <a href="http://ccb.jhu.edu/software/kraken/">kraken</a> installed, so that <code>kraken</code> and
<code>kraken-build</code> are in your path. These are needed if you want to
make your own reference database, or if you use a database to
automatically choose the reference genome.</li>
<li>Optional:<a href="http://blast.ncbi.nlm.nih.gov/Blast.cgi?PAGE_TYPE=BlastDocs&DOC_TYPE=Download">NCBI-blast+</a>.
If this is installed then the QC script will make input files for ACT, to compare the assembly
against a reference.
</li>
</ul>

<p>The QC code is also bundled with the following (they do not need to be installed).</p>

<ul>
<li>Analysis code from the <a href="http://gage.cbcb.umd.edu/index.html">GAGE</a> assembly evaluation
project. We are grateful to the GAGE authors for permission to modify and
redistribute this
code.</li>
<li>
<a href="http://ratt.sourceforge.net/">RATT</a> is used to transfer annotation from a reference
onto the assembly.</li>
</ul>


<hr>
<h2>
<a name="references" class="anchor" href="#references"><span class="octicon octicon-link"></span></a>References</h2>

<p><a href="http://www.nature.com/nmeth/journal/v9/n1/full/nmeth.1814.html"><strong>Adapter sequences:</strong></a> Quail, M. a et al. <em>Optimal enzymes for amplifying sequencing libraries</em>. Nat. Methods 9, 10-1 (2012).</p>

<p><a href="http://genome.cshlp.org/content/early/2012/01/12/gr.131383.111"><strong>GAGE:</strong></a> Salzberg, S. L. et al. <em>GAGE: A critical evaluation of genome
assemblies and assembly algorithms</em>. Genome Res. 22, 557-67 (2012).</p>

<p><a href="http://www.biomedcentral.com/1471-2105/14/160"><strong>KMC:</strong></a> Deorowicz, S., Debudaj-Grabysz, A. &amp; Grabowski, S. <em>Disk-based k-mer
counting on a PC</em>. BMC Bioinformatics 14, 160 (2013).</p>

<p><a href="http://genomebiology.com/2014/15/3/R46"><strong>Kraken:</strong></a> Wood, D. E. &amp; Salzberg, S. L. <em>Kraken: ultrafast metagenomic
sequence classification using exact alignments</em>.
Genome Biol. 15, R46 (2014).</p>

<p><a href="http://genomebiology.com/2004/5/2/r12"><strong>MUMmer:</strong></a> Kurtz, S. et al. <em>Versatile and open software for comparing large
genomes</em>. Genome Biol. 5, R12 (2004).</p>

<p><strong>R:</strong> R Core Team (2013). <em>R: A language and environment for statistical
computing</em>. R Foundation for Statistical Computing, Vienna, Austria.
URL <a href="http://www.R-project.org/">http://www.R-project.org/</a>.</p>

<p><a href="http://nar.oxfordjournals.org/content/39/9/e57"><strong>RATT:</strong></a> Otto, T. D., Dillon, G. P., Degrave, W. S. &amp; Berriman, M.
<em>RATT: Rapid Annotation Transfer Tool</em>. Nucleic Acids Res. 39, e57 (2011).</p>

<p><a href="http://bioinformatics.oxfordjournals.org/content/25/16/2078.abstract"><strong>SAMtools:</strong></a> Li, H. et al. <em>The Sequence Alignment/Map format and SAMtools</em>.
Bioinformatics 25, 2078-9 (2009).</p>

<p><a href="http://bioinformatics.oxfordjournals.org/content/early/2014/04/12/bioinformatics.btu170"><strong>Trimmomatic:</strong></a> Bolger, A. M., Lohse, M. &amp; Usadel, B. <em>Trimmomatic: A
flexible trimmer for Illumina Sequence Data</em>. Bioinformatics 1-7 (2014).</p>
      </section>
    </div>

    <!-- FOOTER  -->
    <div id="footer_wrap" class="outer">
      <footer class="inner">
        <p class="copyright">IVA maintained by <a href="https://github.com/sanger-pathogens">sanger-pathogens</a></p>
        <p>Published with <a href="http://pages.github.com">GitHub Pages</a></p>
      </footer>
    </div>

                  <script type="text/javascript">
            var gaJsHost = (("https:" == document.location.protocol) ? "https://ssl." : "http://www.");
            document.write(unescape("%3Cscript src='" + gaJsHost + "google-analytics.com/ga.js' type='text/javascript'%3E%3C/script%3E"));
          </script>
          <script type="text/javascript">
            try {
              var pageTracker = _gat._getTracker("UA-64654778-2");
            pageTracker._trackPageview();
            } catch(err) {}
          </script>

  </body>
</html>
